### DISCUSSION TITLE
To extract images. To impose a document.

### DISCUSSION THREAD

**Discussion by johnmapeson** [qpdf/qpdf#1260] [Q&A] [answered]
Hello Jay and everyone. A very short question.

- Does QPDF currently provide an option to extract the included images (e.g., if page 17 contains text, then 2x3 inch illustration, and then text again, and you want to extract this illustration as PNG file)?

- And does it currently provide an option to impose a document? "To impose" means that, for example, you have a 12-page document, and you want to convert it to a 6-page document, page 1 of which will be a spread that contains "paper" pages 1 and 2, page 2 will be a spread that contains "paper" pages 3 and 4, and so on. Currently I use pdfjam for this: `pdfjam --nup 2x1 --landscape -o output.pdf -- input.pdf`.

**Comment by jberkenbilt** [comment, marked as answer]
> * Does QPDF currently provide an option to extract the included images (e.g., if page 17 contains text, then 2x3 inch illustration, and then text again, and you want to extract this illustration as PNG file)?

Not in so many words. You can iterate through the images on a page, potentially including images embedded in form XObjects, but you're getting low-level PDF image dictionaries. You'd have to write a lot of code to get them out to PNG files in the general case. ghostscript and poppler both have tools to extract images from PDF.



> * And does it currently provide an option to impose a document?

Not at this time, but it's planned. You could code it using the library, and there is an example of doing overlay which could serve as a skeleton. There is a big effort I've been referring to as the "pages epic" that includes adding support for this, but it hasn't been done and won't be finished for a long time due to lack of resources to build it. There are some other tools built on top of qpdf that can do this, and cups-filters includes some qpdf code to do this when used to print from Linux, but it is not supported "out of the box" by qpdf or exposed from the command-line.

**Comment by johnmapeson** [comment]
Thanks!

> ghostscript and poppler both have tools to extract images from PDF.

And MuPDF: `mutool extract [-r] [-a] [-N] input.pdf`
